src/main/java/ma/hibernate/dao/PhoneDaoImpl.java

```diff
@@ -12,11 +19,43 @@ public PhoneDaoImpl(SessionFactory sessionFactory) {
 
     @Override
     public Phone create(Phone phone) {
-        return null;
+        Session session = null;
+        Transaction transaction = null;
+        try {
+            session = factory.openSession();
+            transaction = session.beginTransaction();
+            session.save(phone);
+            transaction.commit();
+            return phone;
+        } catch (Exception e) {
+            if (transaction != null) {
+                transaction.rollback();
+            }
+            throw new RuntimeException("Can not save phone: " + phone, e);
+        } finally {
+            if (session != null) {
+                session.close();
+            }
+        }
     }
 
     @Override
     public List<Phone> findAll(Map<String, String[]> params) {
-        return null;
+        try (Session session = factory.openSession()) {
+            CriteriaBuilder cb = session.getCriteriaBuilder();
+            CriteriaQuery<Phone> query = cb.createQuery(Phone.class);
+            Root<Phone> phoneRoot = query.from(Phone.class);
+            List<CriteriaBuilder.In<String>> predicates = new ArrayList<>();
+            for (Map.Entry<String, String[]> paramsEntrySet : params.entrySet()) {
+                predicates.add(cb.in(phoneRoot.get(paramsEntrySet.getKey())));
+                for (String parameter : paramsEntrySet.getValue()) {
+                    predicates.get(predicates.size() - 1).value(parameter);
+                }
+            }
+            query.where(cb.and(predicates.toArray(new Predicate[0])));
+            return session.createQuery(query).getResultList();
+        } catch (Exception e) {
+            throw new RuntimeException("Can not get phones by parameters = " + params, e);
```

Comment:
Can we call params.toString()? What will be the output?

Reply:
Output will be smth like this:
{color=[Ljava.lang.String;@1b6924cb, countryManufactured=[Ljava.lang.String;@3a8d467e, maker=[Ljava.lang.String;@141d3d43}
{color=[Ljava.lang.String;@590adb41, countryManufactured=[Ljava.lang.String;@42aae04d, maker=[Ljava.lang.String;@43bf5397}
{color=[Ljava.lang.String;@3679d92e, countryManufactured=[Ljava.lang.String;@47b11ec7, maker=[Ljava.lang.String;@58fa5769}

Reply:
What can we do here to provide an informative message?

Reply:
In about the same way as you did)